clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1154,12 +1155,24 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
         acquire();
         try {
             log.debug("Committing offsets: {} ", offsets);
+            validateOffsets(offsets);
             coordinator.commitOffsetsAsync(new HashMap<>(offsets), callback);
         } finally {
             release();
         }
     }
 
+    private void validateOffsets(Map<TopicPartition, OffsetAndMetadata> offsets) {
+        positions(offsets.keySet());
+        for (Map.Entry<TopicPartition, OffsetAndMetadata> entry : offsets.entrySet()) {
+            OffsetAndMetadata offsetAndMetadata = entry.getValue();
+            Long position = subscriptions.position(entry.getKey());
+            if (offsetAndMetadata.offset() < 0 || offsetAndMetadata.offset() > position) {
```

Comment:
We probably should document this behavior. I am not sure if a larger than position offset is necessarily invalid. I saw sometimes users would want to just reset their offset of a group to a certain position which might be higher than the current position. Arguably this can also be achieved by calling `seek()` then `commit()`, but it maybe a little inconvenient if there are a lot of partitions.